src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/HU2PhaseFlux.hpp

```diff
@@ -0,0 +1,813 @@
+                                                       phaseMob,
+                                                       dPhaseMob,
+                                                       k_up,
+                                                       mob,
+                                                       dMob_dP,
+                                                       dMob_dC );
+      totMob += mob;
+      dTotMob_dP += dMob_dP;
+      for( localIndex jc = 0; jc < numComp; ++jc )
+      {
+        dTotMob_dC[jc] += dMob_dC[jc];
+      }
+    }
+
+    // upwind based on totFlux sign
+    upwindMobility< numComp, numFluxSupportPoints >( ip,
+                                                     seri,
+                                                     sesri,
+                                                     sei,
+                                                     totFlux,
+                                                     phaseMob,
+                                                     dPhaseMob,
+                                                     k_up,
+                                                     mob,
+                                                     dMob_dP,
+                                                     dMob_dC );
+    totMob += mob;
+    dTotMob_dP += dMob_dP;
+    for( localIndex jc = 0; jc < numComp; ++jc )
+    {
+      dTotMob_dC[jc] += dMob_dC[jc];
+    }
+
+    totMob = LvArray::math::max( totMob, minTotMob );
+    real64 const invTotMob = 1 / totMob;
+
+    // fractional flow for viscous part as \lambda_i^{up}/\sum_{NP}(\lambda_j^{up})
+
+    real64 const fractionalFlow = mob * invTotMob;
+    real64 const dFractionalFlow_dP = (dMob_dP - fractionalFlow * dTotMob_dP) * invTotMob;
+    real64 dFractionalFlow_dC[numComp]{};
+    for( localIndex jc = 0; jc < numComp; ++jc )
+    {
+      dFractionalFlow_dC[jc] = (dMob_dC[jc] - fractionalFlow * dTotMob_dC[jc]) * invTotMob;
+    }
+
+    /// Assembling the viscous flux (and derivatives) from fractional flow and total velocity as \phi_{\mu} = f_i^{up,\mu} uT
+
+    real64 const viscousPhaseFlux = fractionalFlow * totFlux;
+    real64 dViscousPhaseFlux_dP[numFluxSupportPoints]{};
+    real64 dViscousPhaseFlux_dC[numFluxSupportPoints][numComp]{};
+
+    // fractionalFlow derivatives
+    dViscousPhaseFlux_dP[k_up] += dFractionalFlow_dP * totFlux;
+    for( localIndex jc = 0; jc < numComp; ++jc )
+    {
+      dViscousPhaseFlux_dC[k_up][jc] += dFractionalFlow_dC[jc] * totFlux;
+    }
+
+    // Ut derivatives
+    for( localIndex ke = 0; ke < numFluxSupportPoints; ++ke )
+    {
+      dViscousPhaseFlux_dP[ke] += fractionalFlow * dTotFlux_dP[ke];
+      for( localIndex jc = 0; jc < numComp; ++jc )
+      {
+        dViscousPhaseFlux_dC[ke][jc] += fractionalFlow * dTotFlux_dC[ke][jc];
+      }
+    }
+
+    // accumulate in the flux and its derivatives (need to be very careful doing that)
+    phaseFlux += viscousPhaseFlux;
```

Comment:
not sure i got it, can you elaborate?

Reply:
I mean for all contribution `viscous`, `capillary` and `gravity`, it will be added to the `phaseFlux` the same way. To avoid copy-paste and possible changes, better refactor everything in a helper, isn't it ?

Reply:
oh ok, sure, will do